iroh-relay/src/server/client_conn.rs

```diff
@@ -317,10 +324,153 @@ impl Actor {
     }
 }
 
+/// Rate limiter for reading from a [`RelayedStream`].
+///
+/// The writes to the sink are not rate limited.
+///
+/// This potentially buffers one frame if the rate limiter does not allows this frame.
+/// While the frame is buffered the undernlying stream is no longer polled.
+#[derive(derive_more::Debug)]
+struct RateLimitedRelayedStream {
+    inner: RelayedStream,
+    limiter: Arc<governor::DefaultDirectRateLimiter>,
+    #[debug("Option<Pin<Box<dyn Future<Output = ()>>>>")]
+    delay: Option<Pin<Box<dyn Future<Output = ()> + Send + Sync>>>,
+    buf: Option<anyhow::Result<Frame>>,
+}
+
+impl RateLimitedRelayedStream {
+    fn new(inner: RelayedStream, limiter: governor::DefaultDirectRateLimiter) -> Self {
+        Self {
+            inner,
+            limiter: Arc::new(limiter),
+            delay: None,
+            buf: None,
+        }
+    }
+}
+
+impl Stream for RateLimitedRelayedStream {
+    type Item = anyhow::Result<Frame>;
+
+    #[instrument(name = "rate_limited_relayed_stream", skip_all)]
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        loop {
+            // If we have a delay installed, we need to await it.
+            if let Some(ref mut wait_fut) = self.delay {
+                tokio::pin!(wait_fut);
```

Comment:
I think you can just write `Pin::new(wait_fut)`, which is a bit easier to understand avoids the macro